internal/testutils/proxy.go

```diff
@@ -0,0 +1,130 @@
+ *
+ */
+
+package testutils
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+)
+
+// ProxyServer represents a test proxy server.
+type ProxyServer struct {
+	lis       net.Listener
+	in        net.Conn            // Connection from the client to the proxy.
+	out       net.Conn            // Connection from the proxy to the backend.
+	onRequest func(*http.Request) // Function to check the request sent to proxy.
+	Addr      string              // Address of the proxy
+}
+
+const defaultTestTimeout = 10 * time.Second
+
+// Stop closes the ProxyServer and its connections to client and server.
+func (p *ProxyServer) stop() {
+	p.lis.Close()
+	if p.in != nil {
+		p.in.Close()
+	}
+	if p.out != nil {
+		p.out.Close()
+	}
+}
+
+func (p *ProxyServer) handleRequest(t *testing.T, in net.Conn, waitForServerHello bool) {
+	req, err := http.ReadRequest(bufio.NewReader(in))
+	if err != nil {
+		t.Errorf("failed to read CONNECT req: %v", err)
+		return
+	}
+	if req.Method != http.MethodConnect {
+		t.Errorf("unexpected Method %q, want %q", req.Method, http.MethodConnect)
+	}
+	p.onRequest(req)
+
+	t.Logf("Dialing to %s", req.URL.Host)
+	out, err := net.Dial("tcp", req.URL.Host)
+	if err != nil {
+		in.Close()
+		t.Logf("failed to dial to server: %v", err)
+		return
+	}
+	out.SetDeadline(time.Now().Add(defaultTestTimeout))
+	resp := http.Response{StatusCode: http.StatusOK, Proto: "HTTP/1.0"}
+	var buf bytes.Buffer
+	resp.Write(&buf)
+
+	if waitForServerHello {
+		// Batch the first message from the server with the http connect
+		// response. This is done to test the cases in which the grpc client has
+		// the response to the connect request and proxied packets from the
+		// destination server when it reads the transport.
+		b := make([]byte, 50)
+		bytesRead, err := out.Read(b)
+		if err != nil {
+			t.Errorf("Got error while reading server hello: %v", err)
+			in.Close()
+			out.Close()
+			return
+		}
+		buf.Write(b[0:bytesRead])
+	}
+	p.in = in
+	p.in.Write(buf.Bytes())
+	p.out = out
+
+	go io.Copy(p.in, p.out)
+	go io.Copy(p.out, p.in)
+}
+
+// HTTPProxy initializes and starts a proxy server, registers a cleanup to
+// stop it, and returns the proxy's listener and helper channels.
+func HTTPProxy(t *testing.T, reqCheck func(*http.Request), waitForServerHello bool) *ProxyServer {
+	t.Helper()
+	pLis, err := LocalTCPListener()
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+
+	p := &ProxyServer{
+		lis:       pLis,
+		onRequest: reqCheck,
+	}
+
+	// Start the proxy server.
+	go func() {
+		for {
+			in, err := p.lis.Accept()
+			if err != nil {
+				return
+			}
+			// p.handleRequest is not invoked in a goroutine because the test
+			// proxy currently supports handling only one connection at a time.
+			p.handleRequest(t, in, waitForServerHello)
+		}
+	}()
+	t.Logf("Started proxy at: %q", pLis.Addr())
+	t.Cleanup(p.stop)
+	p.Addr = fmt.Sprintf("localhost:%d", ParsePort(t, pLis.Addr().String()))
```

Comment:
Also why can't this just be `pLis.Addr().String()`?  There should be a comment explaining why it's doing all this extra stuff to seemingly just convert `localhost:123` into `localhost:123`.